addons/l10n_pt_stock/models/stock_picking.py

```diff
@@ -0,0 +1,351 @@
+    l10n_pt_stock_inalterable_hash = fields.Char(string="Inalterability Hash", readonly=True, copy=False)
+    l10n_pt_inalterable_hash_short = fields.Char(
+        string="Short version of the Portuguese hash",
+        compute='_compute_l10n_pt_stock_inalterable_hash_info',
+    )
+    l10n_pt_stock_inalterable_hash_version = fields.Integer(
+        string="Portuguese hash version",
+        compute='_compute_l10n_pt_stock_inalterable_hash_info',
+    )
+    l10n_pt_stock_atcud = fields.Char(string="Portuguese ATCUD", readonly=True, copy=False)
+    l10n_pt_document_number = fields.Char(
+        string="Unique Document Number",
+        compute='_compute_l10n_pt_document_number', store=True,
+        help="Unique identifier made up of the internal document type code, the series name, "
+             "and the number of the document within the series.",
+    )
+    l10n_pt_hashed_on = fields.Datetime(string="Hashed On", readonly=True)
+    # Document type is used in the template (when printed, documents have to present the document type on each page)
+    l10n_pt_document_type = fields.Selection(
+        selection=AT_SERIES_MOVEMENT_DOCUMENT_TYPES,
+        string="Portuguese Document Type",
+        compute='_compute_l10n_pt_document_type',
+        store=True,
+    )
+    l10n_pt_at_series_id = fields.Many2one(
+        'l10n_pt.at.series',
+        string="AT Series",
+        compute='_compute_l10n_pt_at_series_id',
+        store=True,
+    )
+    l10n_pt_print_version = fields.Selection(
+        selection=[
+            ('original', 'Original'),
+            ('reprint', 'Reprint'),
+        ],
+        string="Version of Printed Document",
+        copy=False,
+    )
+    l10n_pt_start_transport_date = fields.Datetime(
+        'Start of Transport Date',
+        store=True,
+        default=fields.Datetime.now,
+        tracking=True,
+        help="Date and time of start of transport",
+    )
+    l10n_pt_show_no_at_series_warning = fields.Boolean(compute='_compute_l10n_pt_show_no_at_series_warning')
+
+    ####################################
+    # OVERRIDES
+    ####################################
+
+    def write(self, vals):
+        if not vals:
+            return True
+        for picking in self.filtered(lambda p: p.country_code == 'PT'):
+            violated_fields = set(vals).intersection(picking._get_integrity_hash_fields() + ['l10n_pt_stock_inalterable_hash'])
+            if violated_fields and picking.l10n_pt_stock_inalterable_hash:
+                raise UserError(_(
+                    "This document is protected by a hash. "
+                    "Therefore, you cannot edit the following fields: %s.",
+                    ', '.join(f['string'] for f in self.fields_get(violated_fields).values())
+                ))
+        return super().write(vals)
+
+    def button_validate(self):
+        picking = super().button_validate()
```

Comment:
`button_validate` can also return an action or `True`. In case of an action, the pickings state is not `done`